cmd/cmd_test.go

```diff
@@ -42,13 +74,315 @@ func (s *cmdSuite) TestDistroSupportsReExec(c *C) {
+	d := c.MkDir()
+	selfExe := filepath.Join(d, "self-exe")
+	defer cmd.MockVersion("2")()
+	defer cmd.MockSelfExe(selfExe)()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(d, c.MkDir())()
+	c.Assert(os.Symlink("/bin/true", selfExe), IsNil)
+	c.Assert(os.MkdirAll(filepath.Join(d, "/bin/true"), 0755), IsNil)
+	p := d + "/usr/lib/snapd"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(p+"/info", []byte("VERSION=9999"), 0644), IsNil)
+
+	c.Check(cmd.ExecInCoreSnap, PanicMatches, `>exec of "[^"]+/bin/true" in tests<`)
+	c.Check(s.execCalled, Equals, 1)
+	c.Check(s.lastExecArgv0, Equals, filepath.Join(d, "/bin/true"))
+	c.Check(s.lastExecArgv, DeepEquals, os.Args)
+	c.Check(s.lastExecEnvv, testutil.Contains, "SNAP_DID_REEXEC=1")
+}
+
+func (s *cmdSuite) TestExecInCoreSnapBailsNoCoreSupport(c *C) {
+	d := c.MkDir()
+	selfExe := filepath.Join(d, "self-exe")
+	defer cmd.MockVersion("2")()
+	defer cmd.MockSelfExe(selfExe)()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(c.MkDir(), d)()
+	c.Assert(os.Symlink("/bin/true", selfExe), IsNil)
+	c.Assert(os.MkdirAll(filepath.Join(d, "/bin/true"), 0755), IsNil)
+	// no "info" -> no core support:
+	// p := d + "/usr/lib/snapd"
+	// c.Assert(os.MkdirAll(p, 0755), IsNil)
+	// c.Assert(ioutil.WriteFile(p+"/info", []byte("VERSION=9999"), 0644), IsNil)
+
+	cmd.ExecInCoreSnap()
+	c.Check(s.execCalled, Equals, 0)
+}
+
+func (s *cmdSuite) TestExecInCoreSnapMissingExe(c *C) {
+	d := c.MkDir()
+	selfExe := filepath.Join(d, "self-exe")
+	defer cmd.MockVersion("2")()
+	defer cmd.MockSelfExe(selfExe)()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(c.MkDir(), d)()
+	c.Assert(os.Symlink("/bin/true", selfExe), IsNil)
+	// missing exe:
+	// c.Assert(os.MkdirAll(filepath.Join(d, "/bin/true"), 0755), IsNil)
+	p := d + "/usr/lib/snapd"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(p+"/info", []byte("VERSION=9999"), 0644), IsNil)
+
+	cmd.ExecInCoreSnap()
+	c.Check(s.execCalled, Equals, 0)
+}
+
+func (s *cmdSuite) TestExecInCoreSnapBadSelfExe(c *C) {
+	d := c.MkDir()
+	selfExe := filepath.Join(d, "self-exe")
+	defer cmd.MockVersion("2")()
+	defer cmd.MockSelfExe(selfExe)()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(c.MkDir(), d)()
+	// no self/exe -> no re-exec
+	// c.Assert(os.Symlink("/bin/true", selfExe), IsNil)
+	c.Assert(os.MkdirAll(filepath.Join(d, "/bin/true"), 0755), IsNil)
+	p := d + "/usr/lib/snapd"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(p+"/info", []byte("VERSION=9999"), 0644), IsNil)
+
+	cmd.ExecInCoreSnap()
+	c.Check(s.execCalled, Equals, 0)
+}
+
+func (s *cmdSuite) TestExecInCoreSnapBailsNoDistroSupport(c *C) {
```

Comment:
The distro support checks that it's a reexec-supporting distro, and a non-classic distro is unsupporting. It's just the easiest way to make that check fail. You think I should add checks for distro names as well (I already have tests that check that distro support fails if unsupporting, checking distro names in particular)